File: Astromaximum/Classes/utils/AmaxSummaryItem.m
//
//  SummaryItem.m
//  Astromaximum
//
//  Created by admin on 13.11.12.
//  Copyright (c) 2012 S&W Axis. All rights reserved.
//

#import "AmaxSummaryItem.h"


@implementation AmaxSummaryItem
@synthesize mEvents = _mEvents;
@synthesize mKey = _mKey;
@synthesize mEventMode = _mEventMode;
@synthesize mActiveEvent = _activeEvent;

enum {
    EVENT_MODE_NONE = 0,
    EVENT_MODE_CURRENT_TIME = 1,
    EVENT_MODE_CUSTOM_TIME = 2
};

- (id)initWithKey:(AmaxEventType)key events:(NSMutableArray *)events
{
    _mKey = key;
    _mEvents = events;
    _mEventMode = EVENT_MODE_NONE;
    return self;
}

- (int)activeEventPositionWithCustomTime:(long)customTime currentTime:(long)currentTime
{
    int index = 0;
    int prev = -1;
    switch (_mKey) {
		case EV_MOON_MOVE:
			for (AmaxEvent *e in _mEvents) {
				if (e.mEvtype == EV_MOON_MOVE) {
					if (dateBetween(currentTime, [e dateAt:0], [e dateAt:1]) == 0) {
						_mEventMode = EVENT_MODE_CURRENT_TIME;
						return index;
					} else if (dateBetween(customTime, [e dateAt:0], [e dateAt:1]) == 0) {
						_mEventMode = currentTime == 0 ? EVENT_MODE_CUSTOM_TIME : EVENT_MODE_NONE;
						return index;
					}
				}
				++index;
			}
			break;
		case EV_VOC:
		case EV_VIA_COMBUSTA:
			for (AmaxEvent *e in _mEvents) {
				int between = dateBetween(currentTime, [e dateAt:0], [e dateAt:1]);
				if (between == 0) {
					_mEventMode = EVENT_MODE_CURRENT_TIME;
					return index;
				} else if (dateBetween(customTime, [e dateAt:0], [e dateAt:1]) == 0) {
					_mEventMode = currentTime == 0 ? EVENT_MODE_CUSTOM_TIME : EVENT_MODE_NONE;
					return index;
				} else if (between == 1) {
					_mEventMode = EVENT_MODE_NONE;
					return index;
				}
				prev = index;
				++index;
			}
			return prev;
		default:
			for (AmaxEvent *e in _mEvents) {
				if (dateBetween(currentTime, [e dateAt:0], [e dateAt:1]) == 0) {
					_mEventMode = EVENT_MODE_CURRENT_TIME;
					return index;
				} else if (dateBetween(customTime, [e dateAt:0], [e dateAt:1]) == 0) {
					_mEventMode = currentTime == 0 ? EVENT_MODE_CUSTOM_TIME : EVENT_MODE_NONE;
					return index;
				}
				++index;
			}
			break;
    }
    return -1;    
}

- (void)calculateActiveEventWithCustomTime:(long)customTime currentTime:(long)currentTime
{
    int pos = [self activeEventPositionWithCustomTime:customTime currentTime:currentTime];
//    _activeEvent = (pos == -1) ? nil : [_mEvents objectAtIndex:0];
    _activeEvent = [_mEvents count] == 0 ? nil : [_mEvents objectAtIndex:0];
}

@end

File: Astromaximum/Classes/utils/AmaxSummaryItem.swift
//
//  SummaryItem.m
//  Astromaximum
//
//  Created by admin on 13.11.12.
//  Copyright (c) 2012 S&W Axis. All rights reserved.
//

class AmaxSummaryItem {

    var mEvents = [AmaxEvent]()
    var mKey: AmaxEventType
    var mCusSelection = 0

    enum EventMode: Int {
        case none = 0
        case currentTime = 1
        case customTime = 2
    }

    var mEventMode: EventMode

    init(key: AmaxEventType, events: [AmaxEvent]) {
        mKey = key
        mEvents = events
        mEventMode = .none
    }

    func activeEventPosition(customTime: Int, currentTime: Int) -> Int {
        var index = 0
        var prev = -1
        switch mKey {
    		case EV_MOON_MOVE:
    			for e in mEvents {
    				if e.mEvtype == EV_MOON_MOVE {
                        if dateBetween(currentTime, e.date(at: 0), e.date(at: 1)) == 0 {
                            mEventMode = .currentTime
    						return index
    					}
                        else if dateBetween(customTime, e.date(at: 0), e.date(at: 1)) == 0 {
                            mEventMode = currentTime == 0 ? .customTime : .none
    						return index
    					}
    				}
    				index += 1
    			 }
    		case EV_VOC, EV_VIA_COMBUSTA:
    			for e in mEvents {
                    let between = dateBetween(currentTime, e.date(at: 0), e.date(at: 1))
    				if between == 0 {
                        mEventMode = .currentTime
    					return index
    				}
                    else if dateBetween(customTime, e.date(at: 0), e.date(at: 1)) == 0 {
    					mEventMode = currentTime == 0 ? .customTime : .none
    					return index
    				}
                    else if between == 1 {
                        mEventMode = .none
    					return index
    				}
    				prev = index
    				index += 1
    			 }
    			return prev
            case EV_ASP_EXACT:
                for e in mEvents {
                    if currentTime < e.date(at: 0) {
                        return index - 1
                    }
                    else {
                        index += 1
                    }
                }
            case EV_SUN_DAY:
                return index
    		default:
    			for e in mEvents {
    				if dateBetween(currentTime, e.date(at: 0), e.date(at: 1)) == 0 {
                        mEventMode = .currentTime
    					return index
    				}
                    else if dateBetween(customTime, e.date(at: 0), e.date(at: 1)) == 0 {
    					mEventMode = currentTime == 0 ? .customTime : .none
    					return index
    				}
    				index += 1
    			 }
        }
        return -1    
    }

    enum ExtendedRangeMode {
        case none
        case oneDay
        case twoDays
        case month
        case year
    }
    
    func extendedRangeMode() -> ExtendedRangeMode {
        switch mKey {
        case EV_MOON_MOVE:
            return .twoDays
        case EV_PLANET_HOUR, EV_PLANET_HOUR_EXT:
            return .oneDay
        case EV_MOON_SIGN:
            return .month
        case EV_RETROGRADE, EV_ASP_EXACT, EV_SUN_DEGREE, EV_TITHI:
            return .year
        default:
            return .none
        }
    }
    
    func getCusSelEvent() -> AmaxEvent? {
        return mEvents[mCusSelection]
    }

    func recalcSelection(time: Int, isCustom: Bool) {
        return
        /*if (events == null) {
            return;
        }
        if (isCustom) {
            cusSelection = -1;
        } else {
            nowSelection = -1;
        }
//    if(!isCustom /*&& !Summary.isCurrentDay*/) {
//      return;
//    }
        if (type == Event.EV_TATTVAS) {
            long date = events[0].getDate0();
            for (int i = 0; i < Astromaximum.TATTVAS_IN_DAY; ++i) {
                if ((time >= date) && (time < date + Astromaximum.MSECINTATTVA)) {
                    if (isCustom) {
                        cusSelection = i;
                    } else {
                        nowSelection = i;
                    }
                    break;
                }
                date += Astromaximum.MSECINTATTVA;
            }
            return;
        }

        for (int i = 0; i < events.length; i++) {
            boolean flg = false;
            Event ev = events[i];
            if (ev != null) {
                if (type == Event.EV_RISE) {
                    long delta = time - ev.getDate0();
                    flg = (delta > DEGREE_DELTA_MSEC1) && (delta < DEGREE_DELTA_MSEC2);
                }
                else {
                    if (!(type == Event.EV_MOON_MOVE && ev.degree != 200)) {
                        flg = contains(events[i], time);
                    }
                }
                if (!flg) {
                    continue;
                }
                if (isCustom) {
                    cusSelection = i;
                } else {
                    nowSelection = i;
                }
                break;
            }
        }*/
    }
}
